### PR TITLE
feat(tfhe): add zk-pok code base

### PR DESCRIPTION
**Pull request by IceTDrinker**
- integration of work done by Sarah in the repo

### CODE REVIEW

**Comment by IceTDrinker** [comment]
cuda C API tests likely need the ZK pok feature @tmontaigu

**Review by IceTDrinker — COMMENTED** [review]
Some comments that were waiting apparently, will continue this afternoon

File: tfhe/src/integer/server_key/radix_parallel/ilog2.rs [hunk 1]
```diff
@@ -241,6 +241,7 @@ impl ServerKey {
         let counter_num_blocks = ((num_bits_in_ciphertext - 1).ilog2() + 1 + 1)
             .div_ceil(self.message_modulus().0.ilog2()) as usize;
 
+        // 11111000
```
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/ilog2.rs:244]: I was explaining something via zoom and used that file as a whiteboard

File: Makefile [hunk 2]
```diff
@@ -175,6 +175,10 @@ fmt_gpu: install_rs_check_toolchain
 	cargo "$(CARGO_RS_CHECK_TOOLCHAIN)" fmt
 	cd "$(TFHECUDA_SRC)" && ./format_tfhe_cuda_backend.sh
 
+.PHONY: fmt_c_tests # Format c tests
+fmt_c_tests:
+	find tfhe/c_api_tests/ -regex '.*\.\(cpp\|hpp\|cu\|c\|h\)' -exec clang-format -style=file -i {} \;
+
```
[the remark this answers is not in the document]
Comment (tmontaigu) [reply Makefile:181]: no

File: tfhe/src/core_crypto/algorithms/lwe_encryption.rs [hunk 3]
```diff
@@ -1860,61 +2048,178 @@ pub fn encrypt_lwe_ciphertext_with_compact_public_key<
...
+
+    Ok(prove(
+        (public_params, &public_commit),
+        &private_commit,
+        load,
+        random_generator,
+    ))
+}
+
+fn encrypt_lwe_compact_ciphertext_list_with_compact_public_key_impl<
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_encryption.rs:2216]: looks like there is rayon in there so this is the parallel variant ?

File: tfhe/src/core_crypto/algorithms/lwe_encryption.rs [hunk 4]
```diff
@@ -1964,67 +2270,73 @@ pub fn encrypt_lwe_compact_ciphertext_list_with_compact_public_key<
...
-                        output_mask.as_mut(),
-                        mask_noise_distribution,
-                    );
+                let zk_body_convolved = polymul_rev(pk_body.as_ref(), binary_random_slice);
 
-                // Fill the body chunk afterwards manually as it most likely will be smaller than
-                // the full convolution result. rev(b convolved r) + Delta * m + e2
-                // taking noise from Chi_2 for the body part of the encryption
-                // The reverse is to make the first element product match the single ciphertext case
+                assert_eq!(&pk_body_convolved, &zk_body_convolved);
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_encryption.rs:2333]: still some debug asserts here, could you check if others slipped through ?

File: tfhe/src/core_crypto/algorithms/lwe_encryption.rs [hunk 5]
```diff
@@ -1964,67 +2270,73 @@ pub fn encrypt_lwe_compact_ciphertext_list_with_compact_public_key<
...
+
+                slice_wrapping_add_assign(output_mask.as_mut(), mask_noise);
+
+                // Fill the body chunk afterward manually as it most likely will be smaller than
+                // the full convolution result. b convolved r + Delta * m + e2
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_encryption.rs:2338]: careful with the comment and the reverse we did, as we can see above I added a rev at some point in the comment to match the zk order, the comment seems to be missing this ?

File: tfhe/src/core_crypto/algorithms/lwe_encryption.rs [hunk 6]
```diff
@@ -2236,7 +2838,7 @@ pub fn par_encrypt_lwe_compact_ciphertext_list_with_compact_public_key<
                         );
                     },
                     || {
-                        // Fill the temp buffer with b convolved with r
+                        // // Fill the temp buffer with b convolved with r
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_encryption.rs:2841]: unnecessary diff

File: tfhe/src/core_crypto/algorithms/lwe_encryption.rs [hunk 7]
```diff
@@ -2245,17 +2847,15 @@ pub fn par_encrypt_lwe_compact_ciphertext_list_with_compact_public_key<
...
 
-                // Noise from Chi_1 for the mask part of the encryption
-                loop_generator
-                    .unsigned_integer_slice_wrapping_add_random_noise_from_distribution_assign(
-                        output_mask.as_mut(),
-                        mask_noise_distribution,
-                    );
+                let zk_body_convolved = polymul_rev(pk_body.as_ref(), binary_random_slice);
+
+                assert_eq!(&pk_body_convolved, &zk_body_convolved);
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_encryption.rs:2852]: debug can be removed

File: tfhe/src/core_crypto/algorithms/lwe_encryption.rs [hunk 8]
```diff
@@ -2245,17 +2847,15 @@ pub fn par_encrypt_lwe_compact_ciphertext_list_with_compact_public_key<
...
 
-                // Fill the body chunk afterwards manually as it most likely will be smaller than
-                // the full convolution result. rev(b convolved r) + Delta * m + e2
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_encryption.rs:2256]: I believe we should keep this comment here ?

File: tfhe/src/core_crypto/algorithms/lwe_encryption.rs [hunk 9]
```diff
@@ -2264,16 +2864,395 @@ pub fn par_encrypt_lwe_compact_ciphertext_list_with_compact_public_key<
...
+        output,
+        encoded,
+        mask_noise_distribution,
+        body_noise_distribution,
+        secret_generator,
+        encryption_generator,
+    );
+}
+
+/// Encrypt and generates a zero-knowledge proof of an input cleartext list in an output
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_encryption.rs:3007]: parallel variant of

**Review by IceTDrinker — COMMENTED** [review]
in the core part there are some confusions between parallel and non parallel variants it seems, and some debug code is still there, I marked the biggest ones.

Also the private param of the CRS should not be available in the CRS structure otherwise I believe the security of the ZK falls, ping me if you need help

File: tfhe/src/core_crypto/commons/math/random/mod.rs [hunk 10]
```diff
@@ -102,6 +104,86 @@ impl Distribution for UniformTernary {}
...
+                }
+                Bound::Unbounded => {}
+            }
+        }
+
+        true
+    }
+}
+
+impl<T> BoundedDistribution<T, T::Signed> for TUniform<T>
```
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/core_crypto/commons/math/random/mod.rs:155]: No, `Low = High` in the trait definition is just to set Low as by default being the same type as High, we it can be anything else

File: tfhe/src/zk.rs [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+            }
+            Bound::Unbounded => {
+                return Err("requires bounded distribution".into());
+            }
+        };
+
+        let noise_bound = abs_low_bound.max(high_bound);
+
+        if Scalar::BITS > 64 && noise_bound >= (Scalar::ONE << 64usize) {
+            return Err("noise bounds exceeds 64 bits modulus".into());
```
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/zk.rs:100]: yes since they work on 64 data

File: tfhe/src/core_crypto/commons/math/random/mod.rs [hunk 12]
```diff
@@ -102,6 +104,86 @@ impl Distribution for UniformTernary {}
...
+where
+    T: UnsignedInteger,
+{
+    fn low_bound(&self) -> Bound<T::Signed> {
+        Bound::Included(self.min_value_inclusive())
+    }
+
+    fn high_bound(&self) -> Bound<T> {
+        Bound::Included(self.max_value_inclusive().into_unsigned())
+    }
```
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/core_crypto/commons/math/random/mod.rs:165]: for flexibility

File: tfhe/src/core_crypto/algorithms/lwe_encryption.rs (hunk 8 above)
Comment (tmontaigu) [reply tfhe/src/core_crypto/algorithms/lwe_encryption.rs:2256]: The comment is still here I believe

Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/lwe_encryption.rs:2256]: missing the rev(...)

File: tfhe/src/core_crypto/commons/math/random/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/core_crypto/commons/math/random/mod.rs:155]: we can

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply Makefile:181]: https://github.com/zama-ai/tfhe-rs-internal/issues/519

File: tfhe/src/core_crypto/algorithms/lwe_encryption.rs [hunk 13]
```diff
@@ -2264,16 +2832,396 @@ pub fn par_encrypt_lwe_compact_ciphertext_list_with_compact_public_key<
...
+    OutputCont,
+    MaskDistribution,
+    NoiseDistribution,
+    SecretGen,
+    EncryptionGen,
+    G,
+>(
+    lwe_compact_public_key: &LweCompactPublicKey<KeyCont>,
+    output: &mut LweCompactCiphertextList<OutputCont>,
+    messages: &InputCont,
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_encryption.rs:3108]: missed that during review

https://github.com/zama-ai/tfhe-rs-internal/issues/520

don't want to break things now, too late to make small fixes, but I think we should stick to plaintexts

Comment (IceTDrinker) [reply tfhe/src/core_crypto/algorithms/lwe_encryption.rs:3108]: or whatever type + encoding information
